cms/cache/page.py

```diff
@@ -29,41 +34,68 @@ def _page_cache_key(request):
 def set_page_cache(response):
     from django.core.cache import cache
 
-    if not get_cms_setting('PAGE_CACHE'):
-        return response
     request = response._request
-    save_cache = True
-    for placeholder in getattr(request, 'placeholders', []):
-        if not placeholder.cache_placeholder:
-            save_cache = False
-            break
-    if hasattr(request, 'toolbar'):
-        if request.toolbar.edit_mode or request.toolbar.show_toolbar:
-            save_cache = False
-    if request.user.is_authenticated():
-        save_cache = False
-    if not save_cache:
-        add_never_cache_headers(response)
-        return response
-    else:
-        version = _get_cache_version()
-        ttl = get_cms_setting('CACHE_DURATIONS')['content']
-
-        cache.set(
-            _page_cache_key(request),
-            (response.content, response._headers),
-            ttl,
-            version=version
-        )
-        # See note in invalidate_cms_page_cache()
-        _set_cache_version(version)
+
+    if get_cms_setting("PAGE_CACHE") and (
+            not hasattr(request, 'toolbar') or (
+                not request.toolbar.edit_mode and
+                not request.toolbar.show_toolbar and
+                not request.user.is_authenticated()
+            )):
+
+        # This *must* be TZ-aware
+        timestamp = now()
+
+        placeholders = request.current_page.get_placeholders()
```

Comment:
I would add a check above to make sure request.current_page points to an actual page `bool(request.current_page)` because it can be None